presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java

```diff
@@ -0,0 +1,242 @@
+    private final NotificationListener notificationListener = (notification, ignored) -> onNotification(notification);
+    private final SqlTaskManager sqlTaskManager;
+
+    @Inject
+    public GcStatusMonitor(
+            SqlTaskManager sqlTaskManager)
+    {
+        this.sqlTaskManager = requireNonNull(sqlTaskManager, "sqlTaskManager must not be null");
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        for (GarbageCollectorMXBean mbean : ManagementFactory.getGarbageCollectorMXBeans()) {
+            if (mbean.getName().equals("TestingMBeanServer")) {
+                continue;
+            }
+
+            ObjectName objectName = mbean.getObjectName();
+            try {
+                ManagementFactory.getPlatformMBeanServer().addNotificationListener(
+                        objectName,
+                        notificationListener,
+                        null,
+                        null);
+            }
+            catch (JMException e) {
+                throw new RuntimeException("Unable to add GC listener", e);
+            }
+        }
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        for (GarbageCollectorMXBean mbean : ManagementFactory.getGarbageCollectorMXBeans()) {
+            ObjectName objectName = mbean.getObjectName();
+            try {
+                ManagementFactory.getPlatformMBeanServer().removeNotificationListener(objectName, notificationListener);
+            }
+            catch (JMException ignored) {
+            }
+        }
+    }
+
+    private void onNotification(Notification notification)
+    {
+        if (GC_NOTIFICATION_TYPE.equals(notification.getType())) {
+            GarbageCollectionNotificationInfo info = new GarbageCollectionNotificationInfo((CompositeData) notification.getUserData());
+            if (info.isMajorGc()) {
+                onMajorGc();
+            }
+        }
+    }
+
+    private void onMajorGc()
+    {
+        try {
+            logActiveTasks();
+        }
+        catch (Throwable throwable) {
+            log.error(throwable);
+        }
+    }
+
+    private void logActiveTasks()
+    {
+        // We only care about active tasks
+        List<SqlTask> activeSqlTasks = getActiveSqlTasks();
+        ListMultimap<QueryId, SqlTask> activeQueriesToTasksMap = getQueriesToTaskMap(activeSqlTasks);
+        logFormatted(activeQueriesToTasksMap);
+    }
+
+    private static ListMultimap<QueryId, SqlTask> getQueriesToTaskMap(List<SqlTask> sqlTasks)
+    {
+        return sqlTasks.stream()
+                .collect(toImmutableListMultimap(task -> task.getQueryContext().getQueryId(), Function.identity()));
+    }
+
+    private List<SqlTask> getActiveSqlTasks()
+    {
+        return sqlTaskManager.getAllTasks().stream()
+                .filter(task -> !task.getTaskInfo().getTaskStatus().getState().isDone())
+                .collect(toImmutableList());
+    }
+
+    private static void logInfoTable(List<List<String>> table)
```

Comment:
nit: could you please reorder the methods as they are used?

Reply:
Why are you commenting on logInfoTable in particular? This is a method that is defined before it is ever used.

If I take this to it's logical conclusion it means that the methods that are lowest level and least interesting go first at the top of the file? I can reorder the file this way and push the highest level methods to the bottom and the lowest level methods to the top, but this isn't what I see when I look at the code base. Quite the opposite really. 

I don't care at all. I just want clarity so I can get this done.

Reply:
Did the reordering.

Reply:
> Why are you commenting on logInfoTable in particular? This is a method that is defined before it is ever used.

Sorry, it was for no reason. Just a random line in the file.

>  I can reorder the file this way and push the highest level methods to the bottom and the lowest level methods to the top, but this isn't what I see when I look at the code base.

All across the codebase (generally) we are trying to maintain the highest level methods -> lowest level methods.

e.g.: When A calls B and B calls C the methods should be ordered as 

```
A {
B()
}

B {
C()
}

C {
...
}

```

So when you are reading a class upside down - you start with A. If A is not enough, and more details are needed - you will probably be looking for B, and the most intuitive place to find it  - is right below A. And so on ... 


I'm not sure if there's a strong guideline on this. The Google java guideline basically says

> What is important is that each class uses some logical order, which  ...

https://google.github.io/styleguide/javaguide.html#s3.4.2-class-member-ordering

However all across the code generally we were trying to maintain the `highest level methods -> lowest level methods`. 

Perhaps we should update our coding guidelines to reflect this implicit convention

Reply:
Heh, OK that is the ordering I expect. The previous PR had me confused. I'll reorder it.